crates/wit-parser/src/abi.rs

```diff
@@ -148,6 +148,22 @@ impl Resolve {
             params.truncate(0);
             params.push(WasmType::Pointer);
             indirect_params = true;
+        } else {
+            if matches!(
+                (&func.kind, variant),
+                (crate::FunctionKind::Method(_), AbiVariant::GuestExport)
+            ) {
+                // Guest exported methods always receive resource rep as first argument
+                //
+                // TODO: Ideally you would distinguish between imported and exported
+                // resource Handles and then use either I32 or Pointer in abi::push_flat().
+                // But this contextual information isn't available, yet.
+                // See https://github.com/bytecodealliance/wasm-tools/pull/1438 for more details.
+                params.get_mut(0).map(|p| {
+                    assert!(matches!(*p, WasmType::I32));
+                    *p = WasmType::Pointer;
+                });
```

Comment:
Could this instead be:

```rust
assert!(matches!(params[0], WasmType::I32));
params[0] = WasmType::Pointer;
```

to further assert that the first element in the list is defined? (in that it's a bug if the closure here isn't executed)

Reply:
Yes, I think your proposal is more readable. My panic avoidance habits kicked in, but for code generators panics even improve the DX.